File: auth-sdk/src/apikey.ts
import { AxiosInstance } from "axios";
import { TokenManager } from "./token-manager";

export class ApiKey {
  constructor(
    private api: AxiosInstance,
    private tokenManager: TokenManager,
  ) {}

  async list() {
    try {
      const token = await this.tokenManager.getAuthToken();
      const response = await this.api.get(`/apikey`, {
        headers: { Authorization: `Bearer ${token}` },
      });
      return { success: true, data: response.data };
    } catch (error: any) {
      return {
        success: false,
        error: error.response?.data?.message || error.message,
      };
    }
  }

  async create(
    name: string,
    permissions: object = {},
    expiresIn: string | number = -1,
  ) {
    try {
      const token = await this.tokenManager.getAuthToken();
      const response = await this.api.post(
        `/apikey`,
        { name, permissions, expiresIn },
        { headers: { Authorization: `Bearer ${token}` } },
      );
      return { success: true, data: response.data };
    } catch (error: any) {
      return {
        success: false,
        error: error.response?.data?.message || error.message,
      };
    }
  }

  async toggleActive(id: number, active: boolean) {
    try {
      const token = await this.tokenManager.getAuthToken();
      const response = await this.api.patch(
        `/apikey/${id}`,
        { active },
        { headers: { Authorization: `Bearer ${token}` } },
      );
      return { success: true, data: response.data };
    } catch (error: any) {
      return {
        success: false,
        error: error.response?.data?.message || error.message,
      };
    }
  }

  async revoke(id: number) {
    try {
      const token = await this.tokenManager.getAuthToken();
      const response = await this.api.delete(`/apikey/${id}`, {
        headers: { Authorization: `Bearer ${token}` },
      });
      return { success: true, data: response.data };
    } catch (error: any) {
      return {
        success: false,
        error: error.response?.data?.message || error.message,
      };
    }
  }
}

File: auth-sdk/flutter/lib/apikey.dart
import 'account.dart';
import 'config.dart';
import 'token_manager.dart';

class ApiKey {
  ApiKey(this.api, this.tokens);
  final AuthApi api;
  final TokenManager tokens;
  Future<AuthResult<JsonMap>> list() => _call('GET', '/apikey');
  Future<AuthResult<JsonMap>> create(String name,
          [Object permissions = const {}, Object expiresIn = -1]) =>
      _call('POST', '/apikey',
          {'name': name, 'permissions': permissions, 'expiresIn': expiresIn});
  Future<AuthResult<JsonMap>> toggleActive(int id, bool active) =>
      _call('PATCH', '/apikey/$id', {'active': active});
  Future<AuthResult<JsonMap>> revoke(int id) => _call('DELETE', '/apikey/$id');
  Future<AuthResult<JsonMap>> _call(String method, String path,
      [Object? body]) async {
    try {
      return AuthResult.success(await api.send(method, path,
          body: body, headers: authHeaders(await tokens.get())));
    } catch (e) {
      return AuthResult.failure('$e');
    }
  }
}
